### DISCUSSION TITLE
NoSuchMethodError assertOnlyKnownStandardProperties in JDOImplHelper

### DISCUSSION THREAD

**Discussion by fschon** [openmdx/openmdx#41] [Q&A]
I get this runtime exception
Caused by: java.lang.NoSuchMethodError: 'void javax.jdo.spi.JDOImplHelper.assertOnlyKnownStandardProperties(java.util.Map)'

If I change the lines 
`
JDOImplHelper.assertOnlyKnownStandardProperties(overrides);
JDOImplHelper.assertOnlyKnownStandardProperties(props);
`
to
`
implHelper.assertOnlyKnownStandardProperties(overrides); 
implHelper.assertOnlyKnownStandardProperties(props);
`
the runtime exception goes away.

Because I am getting other subsequent errors and the app is not working yet (porting an app based on Opencrx 4.2 to 5.1), I do not know if this is an actual problem or not? I just noticed that elsewhere in the class JDOImplHelper static methods were being called using impllHelper.

**Comment by wfro64** [comment]
I suspect that you are calling the following function:

`JDOHelper.getPersistenceManagerFactory(overrides, "EntityManagerFactory");`

Which properties are you trying to override? A stack trace would also help.

**Comment by fschon** [comment]
Hi Werner,

My setup is that I have a JEE app that runs concurrently with Opencrx and uses the opencrx/mdx jar files directly by including them in the ear/war files.

This is working perfectly for Opencrx 4.2 and I am including the following jar files:

- opencrx-kernel.jar
- opencrx-application.jar
- opencrx-extension.jar
- opencrx-mail.jar
- opencrx-security.jar
- opencrx-resources.jar
- openmdx-base.jar
- openmdx-system.jar
- openmdx-portal.jar
- openmdx-security.jar
- openjpa.jar

I am now porting to Opencrx 5.1 and am currently including the following jars:

- opencrx-config-crx.jar
- opencrx-core.jar
- openmdx-base.jar
- openmdx-portal.jar
- openmdx-security.jar
- openjpa.jar

I am having a lot of difficulties getting the database persistence manager properly connected and the model properly loaded.
For the persistence manager I am calling:
`JDOHelper.getPersistenceManagerFactory("EntityManagerFactory");`
This in turn calls:
`JDOHelper.getPersistenceManagerFactory(overrides, "EntityManagerFactory")`
with null overrides and results in the NoSuchMethodError exception. 

When applying the abovementioned changes from JDOImplHelper... to implHelper... it seems the Persistence Manager is loaded but now I am getting errors loading the models:

> Caused by: org.openmdx.base.exception.ServiceException
        ExceptionDomain = DefaultDomain
        ExceptionCode = PROCESSING_FAILURE
        Description = unknown exception
        Timestamp = 2021-05-27T05:52:09.467Z
        Parameters:
            0:  systemId = xri://+resource/org/openmdx/ui1/xmi1/ui1.xml
            1:  publicId = null
        ... 120 more
Caused by: java.net.MalformedURLException
        ExceptionDomain = DefaultDomain
        ExceptionCode = GENERIC
        Description = unknown protocol: xri
        at java.base/java.net.URL.<init>(URL.java:652)
        at java.base/java.net.URL.<init>(URL.java:541)
        at java.base/java.net.URL.<init>(URL.java:488)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.impl.XMLEntityManager.setupCurrentEntity(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.impl.XMLVersionDetector.determineDocVersion(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
        at deployment.myapp.ear.com.myapp-ui-v10-cdi-1.0-SNAPSHOT.war//org.openmdx.application.xml.spi.ImportHelper.importObjects(ImportHelper.java:244)
        ... 119 more

I am really not sure what to try next?

**Reply by wfro64** [reply]
In order to understand your deployment:

```
apache-tomee-plus-8.0.3/
  + apps/myapp.ear
  + apps/opencrx-core-CRX.ear
```
whereas opencrx-core-CRX.ear is running properly and myapp.ear acts as openCRX client.

**Comment by fschon** [comment]
Hi Werner,

Sorry it took a while to get back ont his. I have just finished another big integration on my system and now have time to circle back to doing this 5.1 upgrade.

My setup is:
```
wildfly/
  + myapp.ear which includes
        opencrx-config-crx.jar
        opencrx-core.jar
        openmdx-base.jar
        openmdx-portal.jar
        openmdx-security.jar
        openjpa.jar
        mywar1.war
        mywar2,war
        ...
```

Opencrx itself is also deployed in a completely separate tomcat container process as a standard installation.

**Reply by wfro64** [reply]
Assuming that mayapp acts as a remote opencrx client: are you able to run https://github.com/opencrx/opencrx-documentation/blob/master/Sdk/RemoteJavaClient.md against your openCRX instance?

**Comment by fschon** [comment]
I run a few utilities using the remote client, but the reason I do not do
this for the main app is largely the significant performance hit of using
the remote client, no ability to call business logic, rereshing the data
cache does not work  and  i was getting spurious NPEs. So instead I pulled
the main opencrx jars out and included them directly in my app. This has
been working great for several years now and also several opencrx version
updates. Only 5.1 has changed the build structure.

**Reply by wfro64** [reply]
Which version of openMDX are you using?

**Reply by fschon** [reply]
ver 2.17.8

**Comment by wfro64** [comment]
https://github.com/opencrx/opencrx-documentation/blob/master/Sdk/NativeJavaClient.md describes how to setup a native openCRX client. Does this work for you?

**Reply by fschon** [reply]
Hi Werner,

Thanks for this. Unfortunately life is never as simple as a minimal example!

I have several differences to the example. Firstly I am deploying about 6 or 7 war files and 2 jar modules in a ear file, and each have hundreds of other dependencies. Secondly I am on Wildfly not Tomcat.

I have officially descended in absolute jar hell with this upgrade. I don't know what structures changed from 4.2 to 5.1 but I cannot find a deployment structure that actually works and from experience the error messages that get thrown in cases like this are generally vague red herrings which mostly do not have much to do with the actual error. These errors vary between linkage errors, duplicate model loading errors, classpath errors etc. Each different deployment structure I try has some or other issue with it.

Currently I have errors that there are java linkage errors because there are different classes for javax.jdo.JDOHelper being found on the various classpaths. I see that openmdx-base actually includes an entire own bundled version of javax.jdo in the jar file and I assume that this is clashing with Wildfly's JDO version, or with the version of the JDO API that Google happen to include in their client libraries, which cannot be separated out or excluded from my project.

I am not sure why openmdx includes a JDO implementation? Is this a stock implementation or is this modified?

The reason I think this might be a red herring is that the opencrx 4.2 also includes a bundled JDO version and I had no problems with that.

**Reply by fschoning** [reply]
Hi Werner,

After some fiddling it turned out that there were a series of interrelated problems, one being that I had compiled opencrx from source and the jars that were created differed somehow from the ones in the release jar. The good news is that I have everything mostly working in 5.1 now.

However I have run into a difference between 4.2 and 5.1 - when calling Contracts.createInvoice(order) the opencrx code will clone the Order and make it an Invoice. In 4.2 this worked fine, however in 5.1 it fails. I suspect that is has to do with the fact that I have some involvedObjects attached to the order that themselves have an involvedObject attached. So a involvedObject is attached to another involvedObject (not to the object it involves, but the involvedObject itself). This is supported by the API but not via the Opencrx GUI.

I don't actually need these involved objects to be copied to the invoice, so I have not checked if 4.2 simply ignored the involvedObjects involvedObjects, but in 5.1 the clone fails with exceptions.

I am trying to debug and trace how I could fix this, and am trying to setup a simple native test client like the one you referenced earlier in this thread, however the SampleNativeClient.java code is now missing from all tags in github.

Could you please:
a) Send me the SampleNativeClient.java code again
b) Maybe give me a pointer on how to fix the cloning problem.

Thanks,
Franz
